src/light_data/light_store.rs

```diff
@@ -132,36 +150,116 @@ impl<F: LurkField> LightStore<F> {
+                    .ok_or_else(|| anyhow!("Invalid u64 pointer: {ptr}"))?;
+                store.insert_scalar_expression(ptr, Some(ScalarExpression::UInt(u.into())));
+            }
+            ExprTag::Str => {
+                store.insert_scalar_expression(ptr, Some(ScalarExpression::Str(String::new())));
+            }
+            ExprTag::Sym => {
+                store.insert_scalar_expression(ptr, Some(ScalarExpression::Sym(Sym::root())));
+            }
+            _ => return Err(anyhow!("Invalid immediate pointer: {ptr}")),
+        };
+        Ok(())
+    }
+
+    fn intern_non_immediate(
+        &self,
+        ptr: ScalarPtr<F>,
+        store: &mut ScalarStore<F>,
+        stack: &mut Vec<ScalarPtr<F>>,
+    ) -> anyhow::Result<()> {
+        match self.get(&ptr) {
+            None => return Err(anyhow!("LightExpr not found for pointer {ptr}")),
+            Some(None) => {
+                // opaque data
+                store.insert_scalar_expression(ptr, None);
+            }
+            Some(Some(expr)) => match (ptr.tag(), expr.clone()) {
+                (ExprTag::Nil, LightExpr::Nil) => {
+                    stack.push(ScalarPtr::from_parts(ExprTag::Sym, *ptr.value()));
+                    store.insert_scalar_expression(ptr, Some(ScalarExpression::Nil));
+                }
+                (ExprTag::Cons, LightExpr::Cons(x, y)) => {
+                    stack.push(x);
+                    stack.push(y);
+                    store.insert_scalar_expression(ptr, Some(ScalarExpression::Cons(x, y)));
                 }
-                Some(LightExpr::Nil) => Some(ScalarExpression::Nil),
-                Some(LightExpr::StrNil) => Some(ScalarExpression::Str(String::from(""))),
-                // TODO: StrCons with opaque contents breaks this
-                Some(LightExpr::StrCons(_, _)) => {
-                    self.insert_scalar_string(*ptr, &mut store)?;
-                    continue;
+                (ExprTag::Comm, LightExpr::Comm(f, x)) => {
+                    stack.push(x);
+                    store.insert_scalar_expression(ptr, Some(ScalarExpression::Comm(f, x)));
                 }
-                Some(LightExpr::SymNil) => Some(ScalarExpression::Sym(Sym::root())),
-                // TODO: SymCons with opaque contents breaks this
-                Some(LightExpr::SymCons(_, _)) => {
-                    self.insert_scalar_symbol(*ptr, &mut store)?;
-                    continue;
+                (ExprTag::Sym, LightExpr::SymCons(_, _)) => {
+                    self.insert_scalar_symbol(ptr, store)?;
                 }
-            };
-            store.insert_scalar_expression(*ptr, se);
+                (ExprTag::Str, LightExpr::StrCons(_, _)) => {
+                    self.insert_scalar_string(ptr, store)?;
+                }
+                (tag, _) => {
+                    return Err(anyhow!(
+                        "Invalid pair of tag and LightExpr: ({tag}, {expr})"
+                    ))
```

Comment:
This hides a lot. In particular, it conflates the case of tags which are simply unsupported, with that in which the wrong kind of `LightExpr` has somehow been associated with a tag.

Given that `LightData` apparently doesn't (yet?) support all Lurk data, it's probably worth separating that case out for clarity. I admit that my perspective is partially formed by my surprise to discover this is the case at all, though. If we make it explicit, then that will be less confusing and will also make it easier to understand the intended behavior.